web/static/components/provenance/state.js

```diff
@@ -0,0 +1,83 @@
+const DEFAULTS = {
+    NODE_SIZE: 10,
+    NODE_COLOR: "#6baed6", // Use CSS default
+    LABEL_SIZE: 14,
+    LABEL_COLOR: "#333333", // Use CSS default
+};
+
+// Observer pattern for state management
+class GraphState {
+    constructor() {
+        this.observers = [];
+        this.state = {
+            nodePositions: {}, // Store node positions
+            nodeStyles: {}, // Store node customizations
+            labelOffsets: {}, // Store label offsets
+            labelStyles: {}, // Store label customizations
+        };
+    }
+
+    addObserver(observer) {
+        this.observers.push(observer);
+    }
+
+    notifyObservers() {
+        this.observers.forEach((observer) => observer.update(this.state));
+    }
```

Comment:
Can you perhaps define how observers might be used with the Graph. I'm trying to think of how the pattern might be useful.

Maybe if each node was tied back to a record, and an update to the relationships would reflect... Just thinking.